shotover/src/transforms/kafka/sink_cluster.rs

```diff
@@ -186,11 +182,197 @@ impl Transform for KafkaSinkCluster {
 }
 
 impl KafkaSinkCluster {
+    fn store_topic(&self, topics: &mut Vec<StrBytes>, topic: TopicName) {
+        if self.topics.get(&topic.0).is_none() && !topics.contains(&topic.0) {
+            topics.push(topic.0);
+        }
+    }
+
+    fn store_group(&self, groups: &mut Vec<StrBytes>, group_id: GroupId) {
+        if self.coordinator_broker_id.get(&group_id.0).is_none() && !groups.contains(&group_id.0) {
+            groups.push(group_id.0);
+        }
+    }
+
+    async fn update_local_nodes(&mut self) {
+        for shared_node in self.nodes_shared.read().await.iter() {
+            let mut found = false;
+            for node in &mut self.nodes {
+                if shared_node.kafka_address == node.kafka_address {
+                    found = true;
+                    node.broker_id = shared_node.broker_id;
+                }
+            }
+            if !found {
+                self.nodes.push(shared_node.clone())
+            }
+        }
+    }
+
     async fn send_requests(
         &mut self,
-        requests: Vec<Message>,
+        mut requests: Vec<Message>,
     ) -> Result<Vec<oneshot::Receiver<Response>>> {
         let mut results = Vec::with_capacity(requests.len());
+        let mut topics = vec![];
+        let mut groups = vec![];
+        for request in &mut requests {
+            match request.frame() {
+                Some(Frame::Kafka(KafkaFrame::Request {
+                    body: RequestBody::Produce(produce),
+                    ..
+                })) => {
+                    for (name, _) in &produce.topic_data {
+                        self.store_topic(&mut topics, name.clone());
+                    }
+                }
+                Some(Frame::Kafka(KafkaFrame::Request {
+                    body: RequestBody::Fetch(fetch),
+                    ..
+                })) => {
+                    for topic in &fetch.topics {
+                        self.store_topic(&mut topics, topic.topic.clone());
+                    }
+                }
+                Some(Frame::Kafka(KafkaFrame::Request {
+                    body: RequestBody::Heartbeat(heartbeat),
+                    ..
+                })) => {
+                    self.store_group(&mut groups, heartbeat.group_id.clone());
```

Comment:
We can join these match arms together since they all do the same thing.

![image](https://github.com/shotover/shotover-proxy/assets/39339036/61f10d61-12e9-4eb8-9660-bab0b7d10c55)

Reply:
oh that is very nice, thanks!

Reply:
Oh nice, even better